### ISSUE TITLE
Using stash() inside a function causes an error, "object 'yyy' not found"

### ISSUE THREAD

**Issue by torfason** [jhrcook/mustashe#20]
I love `mustashe`, its `here()` awareness fits very neatly with my workflow.

However, now I want to use `stash()` inside a function, but code that works in the global environment does not work inside a function. The issue probably has to do with the environment in which the stash code is evaluated, but since I don't see any parameters to change that, I'm filing this as an issue. Below is a full `reprex` illustrating the problem.

A few points:

* Renaming from `xxx` to `yyy` is only done to not get leakage of variable names across scope.
* In my actual use case, I want to set `yyy` based on some logic inside the function and use `depends_on` to trigger a restash if the contents of `yyy` have changed. This results in a different error, but with the same symptoms (works in global environment but not inside a function.
* I'm including in the reprex both versions, with and without `depends_on`.

``` r
library(mustashe)
## clear_stash() # Commented in posted version to protect from inadvertent clearing
#> Clearing stash.
set.seed(42)

xxx <- letters
stash("result_xxx", {sample(xxx,1)}, functional=TRUE ) |> print()
#> Stashing object.
#> [1] "q"

funky <- function() {
  yyy <- letters
  stash("result_yyy", {sample(yyy,1)}, functional=TRUE ) |> print()
}
funky()
#> Stashing object.
#> Error in sample(yyy, 1): object 'yyy' not found

xxxx <- letters
stash("result_xxxx", {sample(xxxx,1)}, functional=TRUE, depends_on="xxxx" ) |> print()
#> Stashing object.
#> [1] "e"

funky2 <- function() {
  yyyy <- letters
  stash("result_yyyy", {sample(yyyy,1)}, functional=TRUE, depends_on="yyyy" ) |> print()
}
funky2()
#> Error in make_hash(depends_on, .TargetEnv): Some dependencies are missing from the environment.
```

<sup>Created on 2022-03-01 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by jhrcook** [comment]
Thank you for pointing this out. It seems like an important and useful feature to add. I'll take a look at your PR this weekend – I'm a bit busy this week.

**Comment by torfason** [comment]
No worries. Note that the PR does not address this issue, but only issue #8.

For this issue, I didn't do a PR or anything, but I looked around and came to the conclusion that this probably had to do with hard-coding the evaluation environment (`.TargetEnv <- .GlobalEnv`). Setting evaluation environments is a pain, but one approach that I know works in another context is the following:

```
eval(parse(text = code),
                envir=new.env(parent=parent.frame(3)))
```

It uses `parent.frame(n)` to walk up the calling stack (the actual value of n needs to be figured out based on how many function calls are between the call to `stash()` and the function where `eval()` is called. Then a `new.env()` is created with the correct environment as a parent, that means that any variables that are used from the environement where `stasth()` was called will "leak through" but any assignments will not inadvertently pollute it.

The final assignment of the result would then not use any `new.env()` but directly assign to `parent.frame(n)`, and the same would probably need to be done for the explicit search for dependencies inside `make_hash()`.

Happy to continue thinking about this whenever you have time to check it out :-)

**Comment by jhrcook** [comment]
Sorry for my above comment on the wrong topic. I do think you are pointing out an important limitation of `stash()` that would be great to address. I think you have a better understanding of the issue than I do, so I first want to make sure that I'm caught up. Please let me know if the following statements are correct:

1. Your second example fails when the code is trying to be evaluated in `evaluate_code()` because `eval()` is using a new environment which is automatically assigned the parent environment using `parent.frame(n = 1)` which will be the environment where the function is called. The variable `yyy` is instead in the environment of the called function.
2. Your last example fails because `make_hash_table()` looks in the `.GlobalEnv` for `yyyy` but it is in the environment of called function.
3. Your proposed solution using `new.env(parent = parent.frame(3))` just travels up the call stack to the user's function that calls `stash()`. As it currently stands (i.e. with `eval(parse(text = code), envir = new.env())`), the parent environment that is implicitly assigned to the new environment is the parent of `evaluate_code()` which is `new_stash()`.

Thank you for your patience on this. It would be great to get this bug figured out, but I want to make sure I understand it first.

**Comment by torfason** [comment]
Hi again, 

I think I agree with your statements, although with the caveat that even if I'm somewhat familiar with environments, thinking about them can make my head hurt 😊.

But seeing that you were so quick to respond, and interested in fixing the issue, I actually decided to just take a stab at implementing the fix. Since environments can be so fickle, I implemented a battery of tests first, before proceeding the fix. After the fix, all the tests pass locally, and they also seem to be passing on GitHub as I'm writing this.

* This PR also fixes the failing tests related to PR #21 issue #8 
* I also updated `test-stash.R`, after the fix, it is no longer necessary to assign input variables to `.GlobalEnv` in the tests, and the cleanup process (removing variables from `.GlobalEnv` was giving warnings because the result was only assigned locally).

Overall, I think this is a pretty complete fix, but I'd be happy to work with you on any additional changes that you think are appropriate.
